vm/src/air_public_input.rs

```diff
@@ -0,0 +1,105 @@
+use std::collections::HashMap;
+
+use felt::Felt252;
+use serde::Serialize;
+use thiserror::Error;
+
+use crate::{
+    types::layout::CairoLayout,
+    vm::{
+        errors::{trace_errors::TraceError, vm_errors::VirtualMachineError},
+        trace::trace_entry::TraceEntry,
+    },
+};
+
+#[derive(Serialize, Debug)]
+pub struct PublicMemoryEntry {
+    address: usize,
+    page: usize,
+    value: Option<Felt252>,
+}
+
+#[derive(Serialize, Debug)]
+pub struct PublicInput<'a> {
+    layout: &'a str,
+    layout_params: Option<&'a CairoLayout>,
+    rc_min: Option<isize>,
+    rc_max: Option<isize>,
+    n_steps: usize,
+    memory_segments: HashMap<&'a str, (usize, Option<usize>)>,
```

Comment:
We could add a #[serde(serialize_with = "path")] attribute to the `memory_segments` field to turn this:
```json
"memory_segments": {
    "range_check_builtin": [
      311638,
      314768
    ],
    "keccak_builtin": [
      622934,
      622942
    ],
    "poseidon_builtin": [
      639318,
      639327
    ],
```
Into something like this:
```json
"memory_segments": {
        "program": {
            "stop_ptr": 5,
            "begin_addr": 1
        },
        "execution": {
            "stop_ptr": 115030,
            "begin_addr": 1689
        },
        "output": {
            "stop_ptr": 115030,
            "begin_addr": 115030
        },
        "pedersen": {
            "stop_ptr": 115030,
            "begin_addr": 115030
        },
```
Without having to store extra strings

Reply:
On the same tone, we should also add a comment on the `memory_segments` field indicating which tuple value is the `begin_addr` and which is the `stop_ptr` to make it more clear to library users

Reply:
Yeah I think we should go the way of cairo-lang and make a `MemoryAddress` struct (or something like that I don't remember the exact name) and replace all tuples with that, I'm not sure how many times tuples are used and if this would go out of the scope of the PR

Reply:
You mean `Relocatable`?

Reply:
I was talking about [this](https://github.com/starkware-libs/cairo-lang/blob/3a32740d0a743af6f4bfe227d1b5ecab0fc2884b/src/starkware/cairo/lang/vm/utils.py#L32C1-L40)

Reply:
Done applying the solution above, but didn't replace all tuple instances with the new struct